icu4c/source/i18n/unicode/messageformat2.h

```diff
@@ -0,0 +1,138 @@
+// © 2016 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#ifndef U_HIDE_DEPRECATED_API
+
+#ifndef MESSAGEFORMAT2_H
+#define MESSAGEFORMAT2_H
+
+#if U_SHOW_CPLUSPLUS_API
+
+/**
+ * \file
+ * \brief C++ API: Formats messages in a language-neutral way using the draft MessageFormat 2.0.
+ */
+
+#if !UCONFIG_NO_FORMATTING
+
+#include "unicode/format.h"
+#include "unicode/parseerr.h"
+#include "unicode/unistr.h"
+#include "unicode/utypes.h"
+
+U_NAMESPACE_BEGIN
+
+/**
+ * <p>MessageFormat2 is a Technical Preview API implementing MessageFormat 2.0.
+ * Since it is not final, documentation has not yet been added everywhere.
+ *
+ * <p>See <a target="github" href="https://github.com/unicode-org/message-format-wg/blob/main/spec/syntax.md">the
+ * description of the syntax with examples and use cases</a> and the corresponding
+ * <a target="github" href="https://github.com/unicode-org/message-format-wg/blob/main/spec/message.abnf">ABNF</a> grammar.</p>
+ *
+ * @internal ICU 74.0 technology preview
+ * @deprecated This API is for technology preview only.
+ */
+
+class U_I18N_API MessageFormat2 : public Format {
```

Comment:
perhaps we should not enshrine the version number in the class name (even for prototyping purposes)?

Reply:
Since there's already a `MessageFormat` class, I wasn't sure what else to do other than naming this class something different; maybe there's some namespace trickery I can use to expose this class as `experimental::MessageFormat` (or something like that) without breaking code that already uses `MessageFormat`? Open to any suggestions here; I don't know what's been done in the past for migration of other parts of the library to a new spec.

Reply:
OK, to match the Java code as closely as possible (in which the Java package name is `message2`, I'm adding a nested namespace `icu::message2` and keeping the class name as-is.